src/cmd/library_list.js

```diff
@@ -0,0 +1,211 @@
+	 *    architectures: String  // comma-separated list of architectures, or * for all
+	 *    filter: String   // text filter to apply to the list of names
+	 * }
+	 * ```
+	 */
+	sections() {
+		throw new Error('not implemented');
+	}
+
+	/**
+	 * Retrieves the default settings:
+	 * {
+	 *    sort: String; default sort order
+	 *    page: Number; default page to retrieve
+	 *    count: Number; default page size
+	 *    name: String; default name filter
+	 *    architectures: String; comma-separated list of architectures to filter on
+	 *    filter: String; string
+	 * }
+	 */
+	settings() {
+		throw Error('not implemented');
+	}
+
+	/**
+	 * The object in which to store the results.
+	 */
+	target() {
+		throw Error('not implemented');
+	}
+
+	apiClient() {
+		throw new Error('not implemented');
+	}
+
+	/**
+	 * Notification of a promise to fetch a given library list.
+	 * @param {Promise} promise   The promise to fetch the list of libraries
+	 * @param {String} name      The configuration name
+	 * @param {Object} settings  The final settings for fetching the list
+	 * @return {Promise|undefined} either a new promise (wrapping the original promise) or
+	 * undefined to use the original unmodified.
+	 */
+	notifyFetchList(promise, name, settings) {
+		return undefined;
+	}
+
+	/**
+	 * Notification that all the lists are being fetched
+	 * @param {Promise} promise   The promise to fetch all library lists
+	 * @param {Object} settings  The settings used to fetch each list
+	 */
+	notifyFetchLists(promise, settings) {
+
+	}
+
+	error(err) {
+		throw err;
+	}
+
+
+}
+
+/**
+ * Implements the library list command.
+ * It populates various library lists with library instances based on the
+ * query provided by the site.
+ */
+export class LibraryListCommand extends Command {
+
+	constructor() {
+		super();
+		this.categories = {
+			official: this._makeCategory('official'),
+			verified: this._makeCategory('verified', 'official'),
+			featured: this._makeCategory('featured'),
+			popular: this._makeCategory('public', 'verified,official'),
+			mine: this._makeCategory('mine', undefined, 'name'),
+			recent: this._makeCategory('all', undefined, '-date')
+		};
+	}
+
+	_makeCategory(scope, excludeScope, sort) {
+		return {scope, excludeScope, sort};
+	}
+
+	/**
+	 * @param {object} state The current conversation state.
+	 * @param {LibraryListCommandSite} site external services.
+	 * @returns {Promise} To run the library list command.
+	 */
+	run(state, site) {
+		let client, settings, sections, config, target;
+
+		let result = Promise.resolve()
+			.then(() => Promise.resolve(site.apiClient()))
+			.then((_client) => {
+				client = _client;
+				return Promise.resolve(site.settings());
+			})
+			.then((_settings) => {
+				settings = _settings;
+				return Promise.resolve(site.sections());
+			})
+			.then((_sections) => {
+				sections = _sections;
+				return Promise.resolve(site.target());
+			})
+			.then((_target) => {
```

Comment:
yep, that's definitely cleaner - `Promise.all()` could be used too. Although the order of execution should be sequential and deterministic, since it may invoke UI for the user.